crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
+    return 1;
+  }
+
+  evp_pkey_set_method(pkey, &pqdsa_asn1_meth);
+
+  PQDSA_KEY *key = PQDSA_KEY_new();
+  if (key == NULL) {
+    // PQDSA_KEY_new sets the appropriate error.
     return 0;
   }
+
+  key->pqdsa = pqdsa;
+  pkey->pkey.pqdsa_key = key;
+
   return 1;
 }
 
-const EVP_PKEY_METHOD dilithium3_pkey_meth = {
-    EVP_PKEY_DILITHIUM3,
-    NULL /* init */,
-    NULL /* copy */,
-    NULL /* cleanup */,
-    pkey_dilithium3_keygen,
-    NULL /* sign_init */,
-    NULL /* sign */,
-    pkey_dilithium3_sign_message,
-    NULL /* verify_init */,
-    NULL /* verify */,
-    pkey_dilithium3_verify_message,
-    NULL /* verify_recover */,
-    NULL /* encrypt */,
-    NULL /* decrypt */,
-    NULL /* derive */,
-    NULL /* paramgen */,
-    NULL /* ctrl */,
-    NULL /* ctrl_str */,
-    NULL /* keygen deterministic */,
-    NULL /* encapsulate deterministic */,
-    NULL /* encapsulate */,
-    NULL /* decapsulate */,
+// Takes an EVP_PKEY_CTX object |ctx| and sets pqdsa parameters defined
+// by |nid|
+int EVP_PKEY_CTX_pqdsa_set_params(EVP_PKEY_CTX *ctx, int nid) {
+  if (ctx == NULL || ctx->data == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  // It's not allowed to change context parameters if
+  // a PKEY is already associated with the context.
+  if (ctx->pkey != NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_OPERATION);
+    return 0;
+  }
+
+  const PQDSA *pqdsa = PQDSA_find_dsa_by_nid(nid);
+  if (pqdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  dctx->pqdsa = pqdsa;
+
+  return 1;
+}
+
+const EVP_PKEY_METHOD pqdsa_pkey_meth = {
+  EVP_PKEY_NISTDSA,
+  pkey_pqdsa_init /* init */,
+  NULL /* copy */,
+  pkey_pqdsa_cleanup /* cleanup */,
+  pkey_pqdsa_keygen,
+  NULL /* sign_init */,
+  NULL /* sign */,
+  pkey_pqdsa_sign_message,
+  NULL /* verify_init */,
+  NULL /* verify */,
+  pkey_pqdsa_verify_message,
+  NULL /* verify_recover */,
+  NULL /* encrypt */,
+  NULL /* decrypt */,
+  NULL /* derive */,
+  NULL /* paramgen */,
+  NULL /* ctrl */,
+  NULL /* ctrl_str */,
+  NULL /* keygen deterministic */,
+  NULL /* encapsulate deterministic */,
+  NULL /* encapsulate */,
+  NULL /* decapsulate */,
 };
+
+DEFINE_LOCAL_DATA(PQDSA_METHOD, sig_ml_dsa_65_method) {
+  out->keygen = ml_dsa_65_keypair;
+  out->sign = ml_dsa_65_sign;
+  out->verify = ml_dsa_65_verify;
+}
+
+DEFINE_LOCAL_DATA(PQDSA, sig_ml_dsa_65) {
+  out->nid = NID_MLDSA65;
+  out->oid = kOIDMLDSA65;
+  out->oid_len = sizeof(kOIDMLDSA65);
+  out->comment = "MLDSA65 ";
+  out->public_key_len = MLDSA65_PUBLIC_KEY_BYTES;
+  out->secret_key_len = MLDSA65_PRIVATE_KEY_BYTES;
+  out->signature_len = MLDSA65_SIGNATURE_BYTES;
+  out->keygen_seed_len = MLDSA65_KEYGEN_SEED_BYTES;
+  out->sign_seed_len = MLDSA65_SIGNATURE_SEED_BYTES;
+  out->method = sig_ml_dsa_65_method();
+}
+
+const PQDSA *PQDSA_find_dsa_by_nid(int nid) {
```

Comment:
This function will need to move with the PQDSA_KEY_xx functions above.

Reply:
Agreed! I put it down here for now, otherwise we get the warning/error `sig_ml_dsa_65()` defined but not used.